src/lib/tls/nss_countbytes.c

```diff
@@ -0,0 +1,244 @@
+    get_counts(fd)->n_written += (n);                      \
+  } while (0)
+
+/** Implementation for PR_Close: frees the 'secret' field, then passes control
+ * to the default close function */
+static PRStatus
+countbytes_close_fn(PRFileDesc *fd)
+{
+  tor_assert(fd);
+
+  tor_nss_bytecounts_t *counts = (tor_nss_bytecounts_t *)fd->secret;
+  tor_free(counts);
+  fd->secret = NULL;
+
+  return default_close_fn(fd);
+}
+
+/** Implementation for PR_Read: Calls the lower-level read function,
+ * and records what it said. */
+static PRInt32
+countbytes_read_fn(PRFileDesc *fd, void *buf, PRInt32 amount)
+{
+  tor_assert(fd);
+  tor_assert(fd->lower);
+
+  PRInt32 result = (fd->lower->methods->read)(fd->lower, buf, amount);
+  if (result > 0)
+    INC_READ(fd, result);
+  return result;
+}
+/** Implementation for PR_Write: Calls the lower-level write function,
+ * and records what it said. */
+static PRInt32
+countbytes_write_fn(PRFileDesc *fd, const void *buf, PRInt32 amount)
+{
+  tor_assert(fd);
+  tor_assert(fd->lower);
+
+  PRInt32 result = (fd->lower->methods->write)(fd->lower, buf, amount);
+  if (result > 0)
+    INC_WRITTEN(fd, result);
+  return result;
+}
+/** Implementation for PR_Writev: Calls the lower-level writev function,
+ * and records what it said. */
+static PRInt32
+countbytes_writev_fn(PRFileDesc *fd, const PRIOVec *iov,
+                     PRInt32 size, PRIntervalTime timeout)
+{
+  tor_assert(fd);
+  tor_assert(fd->lower);
+
+  PRInt32 result = (fd->lower->methods->writev)(fd->lower, iov, size, timeout);
+  if (result > 0)
+    INC_WRITTEN(fd, result);
+  return result;
+}
+/** Implementation for PR_Send: Calls the lower-level send function,
+ * and records what it said. */
+static PRInt32
+countbytes_send_fn(PRFileDesc *fd, const void *buf,
+                   PRInt32 amount, PRIntn flags, PRIntervalTime timeout)
+{
+  tor_assert(fd);
+  tor_assert(fd->lower);
+
+  PRInt32 result = (fd->lower->methods->send)(fd->lower, buf, amount, flags,
+                                              timeout);
+  if (result > 0)
+    INC_WRITTEN(fd, result);
+  return result;
+}
+/** Implementation for PR_Recv: Calls the lower-level recv function,
+ * and records what it said. */
+static PRInt32
+countbytes_recv_fn(PRFileDesc *fd, void *buf, PRInt32 amount,
+                                  PRIntn flags, PRIntervalTime timeout)
+{
+  tor_assert(fd);
+  tor_assert(fd->lower);
+
+  PRInt32 result = (fd->lower->methods->recv)(fd->lower, buf, amount, flags,
+                                              timeout);
+  if (result > 0)
+    INC_READ(fd, result);
+  return result;
+}
+
+/**
+ * Wrap a PRFileDesc from NSPR with a new PRFileDesc that will count the
+ * total number of bytes read and written.  Return the new PRFileDesc.
+ *
+ * This function takes ownership of its input.
+ */
+PRFileDesc *
+tor_wrap_prfiledesc_with_byte_counter(PRFileDesc *stack)
+{
+  if (PREDICT_UNLIKELY(! countbytes_initialized)) {
```

Comment:
It wouldn't be cleaner to explicitly initialize this when Tor starts instead of doing it lazily right before the subsystem is needed?